[FLINK-13407][tests][coordination] Get SlotManager.failUnfulfillableRequest in main thread of StandaloneResourceManager for verification

azagrebin opened the pull request:
## What is the purpose of the change

In `StandaloneResourceManagerTest` we check `isFailingUnfulfillableRequest` in the test thread, although the underlying field `SlotManager.failUnfulfillableRequest` has no synchronisation and set in the main thread of `StandaloneResourceManager`. This can lead to a visibility problem of `SlotManager.failUnfulfillableRequest`: set in the main thread of `StandaloneResourceManager` but never updated in the test thread.

## Brief change log

The idea for the fix is read `SlotManager.failUnfulfillableRequest` in the main thread of `StandaloneResourceManager`, get future result and check in the `StandaloneResourceManagerTest.assertHappensUntil`.

## Verifying this change

run `StandaloneResourceManagerTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

azagrebin commented:
cc @tillrohrmann

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/StandaloneResourceManagerTest.java:
```diff
@@ -105,10 +100,22 @@ private StandaloneResourceManager createResourceManager(Time startupPeriod) thro
+	private static boolean isFailingUnfulfillableRequest(
+			StandaloneResourceManager resourceManager) throws InterruptedException {
+		try {
+			return resourceManager.isFailingUnfulfillableRequestAsync().get();
```
would be simpler to use `CompletableFuture#join` here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/StandaloneResourceManager.java:
```diff
@@ -92,6 +94,11 @@ protected void initialize() throws ResourceManagerException {
+	@VisibleForTesting
+	CompletableFuture<Boolean> isFailingUnfulfillableRequestAsync() {
+		return CompletableFuture.supplyAsync(this::isFailingUnfulfillableRequest, getMainThreadExecutor());
+	}
```
I'd prefer to have these kind of methods in a `TestingStandaloneResourceManager` because then they don't clutter the production code.

azagrebin commented:
Thanks for the review @tillrohrmann I have addressed the comments

I think we can merge this PR separately. I factored out in a separate [JIRA task](https://issues.apache.org/jira/browse/FLINK-13407) as a one step to fix the original [test instability FLINK-13242](https://issues.apache.org/jira/browse/FLINK-13242). Could you assign the task to me?

We can address the latest finding in a separate [JIRA task](https://issues.apache.org/jira/browse/FLINK-13408). Could you assign the task to @xintongsong?
